nbodykit/base/catalogmesh.py

```diff
@@ -340,30 +341,26 @@ def to_real_field(self, out=None, normalize=True):
 
         # read the necessary data (as dask arrays)
         columns = [self.position, self.weight, self.value, self.selection]
-        Position, Weight, Value, Selection = self.read(columns)
-
-        # compute first, so we avoid repeated computes
-        sel = self.base.compute(Selection)
-        Position = Position[sel]
-        Weight = Weight[sel]
-        Value = Value[sel]
 
-        # compute
-        position, weight, value = self.base.compute(Position, Weight, Value)
+        Position, Weight, Value, Selection = self.read(columns)
 
         # ensure the slices are synced, since decomposition is collective
-        N = max(pm.comm.allgather(len(Position)))
+        Nlocalmax = max(pm.comm.allgather(len(Position)))
 
-        # paint data in chunks on each rank
-        chunksize = 1024 ** 2
-        for i in range(0, N, chunksize):
+        # paint data in chunks on each rank;
+        # we do this by chunk 8 million is pretty big anyways.
+        chunksize = _global_options['paint_chunk_size']
+        for i in range(0, Nlocalmax, chunksize):
             s = slice(i, i + chunksize)
 
             if len(Position) != 0:
 
+                # selection has to be computed many times when data is `large`.
+                sel = self.base.compute(Selection[s])
+
                 # be sure to use the source to compute
                 position, weight, value = \
-                    self.base.compute(Position[s], Weight[s], Value[s])
+                    self.base.compute(Position[s][sel], Weight[s][sel], Value[s][sel])
```

Comment:
I think we might want to apply `` sel`` to the arrays after computing to be safer (this was the logic used in 4e41aab0).